lm_eval/tasks/recoder.py

```diff
@@ -0,0 +1,440 @@
+                end_tokens.append(self.get_position_sentinel(noised_start))
+                end_tokens.extend(original_tokens)
+                end_tokens.append(self.sentinel_eos)
+            elif operation.operation == self.EDIT:
+                noised_document.extend(operation.noised_tokens)
+                # EDIT is its own opposite
+                end_tokens.append(self.get_operation_sentinel(self.EDIT))
+                end_tokens.append(self.get_position_sentinel(noised_start))
+                end_tokens.append(self.get_position_sentinel(noised_end))
+                end_tokens.extend(original_tokens)
+                end_tokens.append(self.sentinel_eos)
+            elif operation.operation == self.INSERT:
+                noised_document.extend(operation.noised_tokens)
+                # DELETE is the opposite of the INSERT noising operation
+                end_tokens.append(self.get_operation_sentinel(self.DELETE))
+                end_tokens.append(self.get_position_sentinel(noised_start))
+                end_tokens.append(self.get_position_sentinel(noised_end))
+                end_tokens.extend(original_tokens)
+                end_tokens.append(self.sentinel_eos)
+            all_end_tokens.append(end_tokens)
+        return noised_document, all_end_tokens
+
+    def apply_noise_to_parts(self, noised_part, all_end_parts):
+        end_parts = [t for et in all_end_parts for t in et]
+        return (noised_part + [self.noised_eos] + end_parts)[:self.tokens_per_sample]
+
+    def apply_noise(self, noise_document: NoisedDocument):
+        noised_part, all_end_parts = self.build_parts(noise_document)
+        return self.apply_noise_to_parts(noised_part, all_end_parts)
+
+class RecoderPython512(Recoder):
+    FILENAME = "/checkpoint/dpf/data/code_corpus_noised/512-512_mr-2_mdr-2/val_github_python_forkless_open-source_2star+_0.jsonl"
+
+def make_tokenizer(tokens_per_sample):
+    from transformers import AutoTokenizer
+    tokenizer = AutoTokenizer.from_pretrained("facebook/incoder-1B")
+
+    tokenizer.add_tokens(Recoder.NOISED_EOS_TOKEN)
+    noised_eos = tokenizer.vocab[Recoder.NOISED_EOS_TOKEN]
+
+    sentinel_token_ids_per_operation = {}
+    for operation, token_str in [
+        (Recoder.DELETE, Recoder.DELETE_TOKEN),
+        (Recoder.EDIT, Recoder.EDIT_TOKEN),
+        (Recoder.INSERT, Recoder.INSERT_TOKEN),
+    ]:
+        tokenizer.add_tokens(token_str, special_tokens=True)
+        sentinel_token_ids_per_operation[operation] = tokenizer.vocab[token_str]
+
+    sentinel_eos = tokenizer.vocab["<|endofmask|>"]
+
+    position_token_ids = []
+    for i in range(tokens_per_sample+1):
+        token_str = Recoder.make_position(i)
+        tokenizer.add_tokens(token_str, special_tokens=True)
+        position_token_ids.append(tokenizer.vocab[token_str])
+    return tokenizer, noised_eos, sentinel_eos, sentinel_token_ids_per_operation, position_token_ids
```

Comment:
qq, why this task is called "recoder", while ppl based ones are called "incoder"?